File: solutions/javascript/luhn/1/luhn.js
export const valid = (number) => {
  let sum = 0;
  let odd = true;
  let digitCount = 0;
  for (let i = number.length - 1; i >= 0; i--) {
    const currentChar = number[i];
    if (currentChar === ' ') continue;
    if (+currentChar === NaN) return false;

    sum += odd
      ? +currentChar
      : +currentChar < 5
      ? +currentChar * 2
      : +currentChar * 2 - 9;

    odd = !odd;
    digitCount++;
  }

  if (digitCount < 2) return false;

  return sum % 10 === 0;
};

File: solutions/r/luhn/1/luhn.R
# Determine whether the number is valid.
is_valid <- function(input) {
  cleaned_input <- gsub("\\s", "", input)

  if (!grepl("^\\d\\d+$", cleaned_input)) return(FALSE)
  
  # reverse, multiply, correct, sum
  reversed <- cleaned_input |>
    strsplit("") |>
    unlist() |>
    rev() |>
    as.numeric()
  
  reversed <- reversed * c(1, 2)
  reversed[reversed > 9] <- reversed[reversed > 9] - 9
  sum(reversed) %% 10 == 0
}